include/cuda/functional

```diff
@@ -71,41 +71,41 @@ class __return_type_wrapper {
  public:
   __return_type_wrapper() = delete;
 
-  _LIBCUDACXX_CONSTEXPR_AFTER_CXX11 
+  _LIBCUDACXX_CONSTEXPR_AFTER_CXX11
   __return_type_wrapper(__return_type_wrapper const&) = default;
 
   template <class _Fn,
             class = _CUDA_VSTD::__enable_if_t<_CUDA_VSTD::is_same<_CUDA_VSTD::__decay_t<_Fn>, _DecayFn>::value>>
-  _LIBCUDACXX_CONSTEXPR_AFTER_CXX11 _LIBCUDACXX_INLINE_VISIBILITY 
+  _LIBCUDACXX_CONSTEXPR_AFTER_CXX11 _LIBCUDACXX_INLINE_VISIBILITY
   explicit __return_type_wrapper(_Fn &&__fn) noexcept
     : __fn_(_CUDA_VSTD::forward<_Fn>(__fn)) {}
 
   template <class... _As>
   _LIBCUDACXX_CONSTEXPR_AFTER_CXX11 _LIBCUDACXX_INLINE_VISIBILITY _Ret
   operator()(_As&&... __as) & noexcept {
-#if !defined(__NVCC__) || defined(__CUDA_ARCH__)
-    static_assert(
-        _CUDA_VSTD::is_same<
-            _Ret,
-            typename _CUDA_VSTD::__invoke_of<_DecayFn, _As...>::type
-          >::value,
-        "Return type shall match the proclaimed one exactly");
-#endif
+    NV_IF_TARGET(NV_IS_DEVICE,(
+      static_assert(
+          _CUDA_VSTD::is_same<
+              _Ret,
+              typename _CUDA_VSTD::__invoke_of<_DecayFn, _As...>::type
+            >::value,
+          "Return type shall match the proclaimed one exactly");
+    ))
```

Comment:
I am pretty sure this is not going to work. `NV_IF_TARGET` isn't actually `constexpr`, so the `static_assert` will always fire. I suspect this will need to keep `__CUDA_ARCH__`.

Reply:
I just checked and it actually just eats the whole block:
```cpp

#define _NV_IF_0(t, f) f
#define _NV_IF_1(t, f) t

#define _NV_IF_BIT(b) _NV_EVAL(_NV_IF_##b)
#define _NV_IF__EVAL(fn, t, f) _NV_EVAL(fn(t, f))
#define _NV_IF_EVAL(cond, t, f) _NV_IF__EVAL(_NV_IF_BIT(cond), t, f)
```

Reply:
I tried this and confirmed it does not work this way. The `static_assert` will be triggered in host code:

https://godbolt.org/z/WfTYqYhzd

Reply:
Yeah, this being such an NVCC specific thing, it might be okay to excuse the `__CUDA_ARCH__` if it's not circumventable.